ui/src/map/layers/highlight/publication-candidate-layer.ts

```diff
@@ -0,0 +1,798 @@
+    const end = pointRange?.indexRange.max || lastIndex(alignmentPoints);
+    const points = alignmentPoints.slice(
+        start,
+        end + 1, // +1 for inclusive slicing
+    );
+    const rangeLengthInMeters = expectDefined(last(points)).m - expectDefined(first(points)).m;
+    const rangeLengthInPixels = rangeLengthInMeters / metersPerPixel;
+    const style = new Style({
+        stroke: new Stroke({
+            color: getHighlightColor(candidate.stage, changeType, candidate.operation),
+            width:
+                candidate.stage === PublicationStage.UNSTAGED
+                    ? UNSTAGED_ALIGNMENT_HIGHLIGHT_WIDTH
+                    : STAGED_ALIGNMENT_HIGHLIGHT_WIDTH,
+            lineCap: rangeLengthInPixels < 15 ? 'square' : 'butt',
+        }),
+        zIndex: getHighlightZIndex(
+            candidate.operation,
+            candidate.type,
+            candidate.stage,
+            changeType,
+        ),
+    });
+
+    const feature = new Feature({
+        geometry: new LineString(points.map(pointToCoords)),
+    });
+    feature.setStyle(style);
+    feature.set(dataPropertyByType(type), candidate);
+
+    return feature;
+}
+
+function createLocationTrackCandidateFeatures(
+    candidates: LocationTrackCandidateAndAlignment[],
+    metersPerPixel: number,
+): Feature<LineString>[] {
+    return candidates.flatMap((candidate) =>
+        splitByRanges(candidate.alignment.points, candidate.publishCandidate.geometryChanges).map(
+            (pointRange) =>
+                createAlignmentLineStringFeature(
+                    pointRange,
+                    candidate.alignment.points,
+                    candidate.publishCandidate,
+                    metersPerPixel,
+                    DraftChangeType.LOCATION_TRACK,
+                ),
+        ),
+    );
+}
+
+function createReferenceLineCandidateFeatures(
+    candidates: ReferenceLineCandidateAndAlignment[],
+    metersPerPixel: number,
+): Feature<LineString>[] {
+    return candidates.flatMap((candidate) =>
+        splitByRanges(candidate.alignment.points, candidate.publishCandidate.geometryChanges).map(
+            (pointRange) =>
+                createAlignmentLineStringFeature(
+                    pointRange,
+                    candidate.alignment.points,
+                    candidate.publishCandidate,
+                    metersPerPixel,
+                    DraftChangeType.REFERENCE_LINE,
+                ),
+        ),
+    );
+}
+
+function createTrackNumberCandidateFeatures(
+    candidates: TrackNumberCandidateAndAlignment[],
+    metersPerPixel: number,
+): Feature<LineString>[] {
+    return candidates.flatMap((candidate) =>
+        createAlignmentLineStringFeature(
+            undefined,
+            candidate.alignment.points,
+            candidate.publishCandidate,
+            metersPerPixel,
+            DraftChangeType.TRACK_NUMBER,
+        ),
+    );
+}
+
+const createOfficialLocationTrackFeatures = (
```

Comment:
Tässä käytännössä alkuperäisen geometrian piirto sijaintiraiteelle. Note to self: Voisi olla joku parempikin termi kuin "official"